cmd-dump.c

```diff
@@ -1377,6 +1395,13 @@ static void do_dump_replay(struct uftrace_dump_ops *ops, struct opts *opts,
 		dump_replay_task(ops, task);
 
 		task->timestamp_last = frs->time;
+
+		if (frs->type == UFTRACE_EVENT) {
+			struct uftrace_perf_reader perf;
+			perf.tid = task->tid;
+			strncpy(perf.u.comm.comm, task->args.data, 16);
+			ops->perf_event(ops, &perf, frs);
```

Comment:
Hmm... you cannot know whether it's a perf event.  And you'd better getting a proper perf reader from the file handle.  I think we need to refactor the code to handle events in `do_dump_replay()` first.

Reply:
As you mentioned, I didn't handle the case if the event is not perf event. I wanted to write the code like `do_dump_file()` but the structure of code is different so added here. Do you have any idea on the refactoring of `do_dump_replay()`?

Reply:
I've uploaded some change to support event types for `do_dump_replay()`.  You can handle the `PERF_COMM` event in the `perf_event` callback.  Please check out the [check/dump-event](https://github.com/namhyung/uftrace/compare/check/dump-event) branch.

Reply:
Hmm.. It seems that there's something wrong in the patch fe04d3f. It makes `tests/t-taskname` crash in the below point.
```
1820│ static int __read_rstack(struct ftrace_file_handle *handle,
1821│                          struct ftrace_task_handle **taskp,
1822│                          bool consume)
1823│ {
    ...
1873│         switch (source) {
1874│         case USER:
1875│                 utask->rstack = &utask->ustack;
1876│                 task = utask;
1877│                 break;
1878│         case KERNEL:
1879│                 ktask->rstack = get_kernel_record(kernel, ktask, k);
1880│                 task = ktask;
1881│                 break;
1882│
1883│         case PERF:
1884│                 task = get_task_handle(handle, perf->tid);
1885│                 task->rstack = get_perf_record(handle, perf);
1886│
1887├───────────────> if (task->rstack->addr == EVENT_ID_PERF_COMM) {
1888│                         /* abuse task->args */
1889│                         task->args.data = xstrdup(perf->u.comm.comm);
1890│                         task->args.len  = strlen(perf->u.comm.comm);
1891│                 }
1892│                 break;

(gdb) r                                                                                                                                                           [52/114]
Starting program: /home/honggyu/work/uftrace/uftrace replay --no-pager
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
# DURATION     TID     FUNCTION
   1.213 us [109736] | __monstartup();
   1.230 us [109736] | __cxa_atexit();
            [109736] | main() {
            [109736] |   task_name1() {
            [109736] |     prctl() {
   9.061 us [109736] |     } /* prctl */
  10.231 us [109736] |   } /* task_name1 */
            [109736] |   task_name2() {
   1.203 us [109736] |     pthread_self();
            [109736] |     pthread_setname_np() {
   5.194 us [109736] |     } /* pthread_setname_np */
   7.704 us [109736] |   } /* task_name2 */
  19.028 us [109736] | } /* main */

Program received signal SIGSEGV, Segmentation fault.
0x00000000004349c9 in __read_rstack (handle=0x7fffffffe080, taskp=0x7fffffffe060, consume=true) at /home/honggyu/work/uftrace/utils/fstack.c:1887

(gdb) p task->rstack
$4 = (struct uftrace_record *) 0x0
(gdb)
```

Reply:
There was a mistake..  I've uploaded the branch with the following fix.

```diff
diff --git a/cmd-dump.c b/cmd-dump.c
index e8a1397..8974edf 100644
--- a/cmd-dump.c
+++ b/cmd-dump.c
@@ -1327,6 +1327,9 @@ static void do_dump_file(struct uftrace_dump_ops *ops, struct opts *opts,
                                break;

                        ops->perf_event(ops, perf, rec);
+
+                       /* for re-read perf data from file */
+                       perf->valid = false;
                }
        }

diff --git a/utils/perf.c b/utils/perf.c
index 67999c0..83a7af6 100644
--- a/utils/perf.c
+++ b/utils/perf.c
@@ -455,8 +455,10 @@ struct uftrace_record * get_perf_record(struct ftrace_file_handle *handle,
 {
        static struct uftrace_record rec;

-       if (read_perf_event(handle, perf) < 0)
-               return NULL;
+       if (!perf->valid) {
+               if (read_perf_event(handle, perf) < 0)
+                       return NULL;
+       }

        rec.type  = UFTRACE_EVENT;
        rec.time  = perf->time;
```